### ISSUE TITLE
Allow overriding image / environment specs at flow run time

### ISSUE THREAD

**Issue by bnaul** [PrefectHQ/prefect#3456]
Not sure if this is already in the works with #3333 and other related changes to environments, but I thought this warranted its own issue since it seems like it would be a common request.

From what I understand, environments like `DaskKubernetesEnvironment` (and I assume `DaskCloudProviderEnvironment`) allow you to specify the cluster configuration (Docker image tag, number of workers, etc.) as part of the environment object, but after registering the flow you can no longer update it. I thought maybe passing variables like the image and number of workers to the flow run context would allow a custom environment class to override these when starting a new cluster, but @cicdw pointed out that these context variables don't get propagated until https://github.com/PrefectHQ/prefect/blob/master/src/prefect/engine/cloud/flow_runner.py#L376 which is too late to affect the creation of the cluster.

It seems like there are a couple of options:
1) Populate the `context` earlier so my hacky approach works
2) Extend the new `run_config` to control more than just the `prefect-job` and allow overriding its attributes at runtime
3) This is an anti-pattern for some reason and one should always re-register a flow in order to change how it runs

Obviously re-registering a flow is not too much of a burden, the main motivations for not wanting to are just simpler bookkeeping in the UI and also some race conditions that arise from two people trying to register+run at the same time. So if there's a way around it I do think there would be some benefit.

**Comment by jcrist** [comment]
I think this will all be fixed as the changes from #3333 propagate out, but to check, let me describe the current end design.

The current plan is:
- `Environment` classes are all deprecated, in favor of a new `RunConfig`. Each agent gets one corresponding `RunConfig`, currently only `KubernetesRun` is implemented, but we'll add others eventually.
- Each `RunConfig` class is for configuring the properties of the "job" launched by the agent, but handles nothing more than that. In particular, this means there is no dask-aware `RunConfig` thing.
- Executors are configurable on the `Flow` itself. So to configure your dask executor, you'd set a `DaskExecutor` on the flow itself. `DaskExecutor` is already configurable to use dask-kubernetes (or any other backing cluster class), so this is already doable.
- For the `KubernetesRun` case, the registered config can point to a job file (currently stored in the agent environment, or in `s3` or `gcs`). So if you want to update the job spec after registration, you can register a flow to use a job spec stored externally, and then update that file as needed.

For k8s, most of this is already implemented. For example:

```python
from prefect.run_configs import KubernetesRun

# Use a job template stored in S3 to configure the original job
flow.run_config = KubernetesRun(job_template_path="s3://bucket/path/to/job.yaml")

# Use a DaskExecutor with dask-kubernetes
flow.executor = DaskExecutor(
    cluster_class="dask_kubernetes.KubeCluster",
    cluster_kwargs={...},
    adapt_kwargs={"maximum": 10}
)

# Store the flow in S3
flow.storage = S3(...)

flow.register()
```

The `run_config` stuff is still experimental, so the interfaces might change without notice for a few releases, but at some point we expect to push forward with this plan.

**Comment by jcrist** [comment]
I guess in the above I'm interpreting the request as "some way to update a run configuration without re-registering the flow", which in this case I'd argue the above should work.

If the request is "have some way to configure a single flow run in the UI, by providing a context" then that's not really handled. I could see an argument for extending the context to let it configure the runtime executor, but I'd rather not extend that to also configuring the job started by the agent (at least not via the `context`). So you could use context variables to configure the number of workers, but not properties of the job started by the agent. If possible, I'd like to avoid this complication though.

**Comment by bnaul** [comment]
Thanks @jcrist! Definitely not attached to the context approach, that was just one way I'd explored to get what I wanted within the current framework.

In my case I'm not concerned with customizing the k8s job at all, https://github.com/PrefectHQ/prefect/pull/3459 was basically the only change I had made in my custom YAML. And the separation between job+Dask is definitely more intuitive. But for overriding e.g. `executor.cluster_kwargs` or `adapt_kwargs` at runtime, I guess I'm still missing how that would work without re-registering. In that case `context` does seem like the most directly way to inject new parameters, but maybe there's another approach that would work as well?

**Comment by jcrist** [comment]
With that as the goal, then using the `context` would be the most straightforward solution, since we already expose configuring the context for a flow run. The trick here is that the executor has already been configured statically on the flow, so we'd need to think of a new convention for letting the config update before starting the executor. I think this would likely be specific to the executor itself, and might follow a few patterns. Using the `DaskExecutor` as an example:

- We might allow templating values in `cluster_kwargs`, `client_kwargs`, and `adapt_kwargs` based on the context? We might want to recurse through dicts, lists in the values as well, to support nested templating. I suspect this would cover most cases, but wouldn't be able to dynamically add new config keys, only modify the values.

```python
flow.executor = DaskExecutor(
    cluster_class="dask_kubernetes.KubeCluster",
    cluster_kwargs={"image": "${prefect.context.image}"}
)
```
- We might let `cluster_kwargs` etc... be a callable that returns keyword arguments, so you could dynamically change these as needed? 

```python
flow.executor = DaskExecutor(
    cluster_class="dask_kubernetes.KubeCluster",
    cluster_kwargs=lambda: {"image": prefect.context.image}
)
```

- We don't add a new mechanism. `cluster_class` can already be any callable, so you can do this all yourself as needed (if you wanted to try this out, I'd be curious about your experience).

```python
def make_cluster():
    cluster =  dask_kubernetes.KubeCluster(
        image=prefect.context.image
    )
    cluster.adapt(...)
    return cluster

flow.executor = DaskExecutor(cluster_class=make_cluster)
```

Thoughts? Right now I'm leaning towards doing nothing and going with the last approach here (which should already work), but the first option also seems fine. Not as enthused about option 2.

**Comment by bnaul** [comment]
The last option is basically what I'm trying myself right now. It didn't work before when I was trying to do that templating inside the Environment setup, but by the time the executor creates the cluster that info should be present, right? So actually it does seem like the #3333 pattern has already resolved my issue, just wasn't obvious to me from the start 🙂 

Will leave it to you whether you want to leave this open or close; I do think at least including an example in a doc somewhere that this is possible would be a good step.

**Comment by jcrist** [comment]
Definitely. Please let me know if things aren't working for you. The work in #3333 is still considered experimental/secret (shhhhh), so I think I'll hold off on a doc for now, but will keep this issue open to track it. Thanks for the feature request, happy it seems to already exist :).

**Comment by jcrist** [comment]
> Extend the new run_config to control more than just the prefect-job and allow overriding its attributes at runtime

We're planning on doing this in #3858, closing and tracking the issue there.
